// ionic-angular/umd/index.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports", '@angular/core', '@angular/common', '@angular/platform-browser', '@angular/forms', '@angular/http', './components/action-sheet/action-sheet', './components/alert/alert', './components/app/app', './components/app/app-root', './config/config', './navigation/deep-linker', './platform/dom-controller', './util/events', './util/form', './gestures/gesture-controller', './tap-click/haptic', './gestures/gesture-config', './platform/keyboard', './components/loading/loading', './components/menu/menu-controller', './components/modal/modal', './components/picker/picker', './platform/platform', './platform/platform-registry', './components/popover/popover', './tap-click/tap-click', './components/toast/toast', './config/mode-registry', './transitions/transition-registry', './transitions/transition-controller', './navigation/url-serializer', './components/action-sheet/action-sheet-component', './components/alert/alert-component', './components/app/app-root', './components/loading/loading-component', './components/modal/modal-component', './components/picker/picker-component', './components/popover/popover-component', './components/toast/toast-component', './components/avatar/avatar', './components/backdrop/backdrop', './components/badge/badge', './components/button/button', './components/card/card', './components/checkbox/checkbox', './components/chip/chip', './util/click-block', './components/content/content', './components/datetime/datetime', './components/fab/fab', './components/grid/col', './components/grid/grid', './components/grid/row', './components/icon/icon', './components/img/img', './components/infinite-scroll/infinite-scroll', './components/infinite-scroll/infinite-scroll-content', './components/item/item', './components/item/item-reorder', './components/item/item-sliding', './components/label/label', './components/list/list', './components/list/list-header', './components/menu/menu', './components/menu/menu-close', './components/menu/menu-toggle', './components/input/native-input', './components/nav/nav', './components/nav/nav-pop', './components/nav/nav-push', './components/navbar/navbar', './components/note/note', './components/option/option', './components/nav/overlay-portal', './components/picker/picker-component', './components/radio/radio-button', './components/radio/radio-group', './components/range/range', './components/range/range-knob', './components/refresher/refresher', './components/refresher/refresher-content', './components/scroll/scroll', './components/searchbar/searchbar', './components/segment/segment', './components/select/select', './components/show-hide-when/show-hide-when', './components/slides/slide', './components/slides/slides', './components/spinner/spinner', './components/tabs/tab', './components/tabs/tabs', './components/tabs/tab-button', './components/tabs/tab-highlight', './components/input/input', './components/thumbnail/thumbnail', './components/toggle/toggle', './components/toolbar/toolbar', './components/toolbar/toolbar-item', './components/toolbar/toolbar-title', './components/typography/typography', './components/virtual-scroll/virtual-scroll', './components/virtual-scroll/virtual-item', './components/action-sheet/action-sheet', './components/alert/alert', './components/app/app', './components/avatar/avatar', './components/backdrop/backdrop', './components/badge/badge', './components/button/button', './components/card/card', './components/checkbox/checkbox', './components/chip/chip', './util/click-block', './components/content/content', './components/datetime/datetime', './components/fab/fab', './components/grid/col', './components/grid/grid', './components/grid/row', './components/ion', './components/icon/icon', './components/img/img', './components/infinite-scroll/infinite-scroll', './components/infinite-scroll/infinite-scroll-content', './components/input/input', './components/app/app-root', './components/item/item', './components/item/item-reorder', './components/item/item-sliding', './components/label/label', './components/list/list', './components/list/list-header', './components/loading/loading', './components/menu/menu', './components/menu/menu-close', './components/menu/menu-controller', './components/menu/menu-toggle', './components/menu/menu-types', './components/modal/modal', './components/nav/nav', './components/nav/nav-pop', './components/nav/nav-push', './components/navbar/navbar', './components/input/native-input', './components/note/note', './components/option/option', './components/nav/overlay-portal', './components/picker/picker', './components/popover/popover', './components/radio/radio-button', './components/radio/radio-group', './components/range/range', './components/range/range-knob', './components/refresher/refresher', './components/refresher/refresher-content', './components/scroll/scroll', './components/searchbar/searchbar', './components/segment/segment', './components/select/select', './components/show-hide-when/show-hide-when', './components/slides/slide', './components/slides/slides', './components/spinner/spinner', './components/tabs/tab', './components/tabs/tab-button', './components/tabs/tab-highlight', './components/tabs/tabs', './tap-click/tap-click', './components/toast/toast', './components/toggle/toggle', './components/toolbar/toolbar', './components/toolbar/toolbar-item', './components/toolbar/toolbar-title', './components/thumbnail/thumbnail', './components/typography/typography', './components/virtual-scroll/virtual-scroll', './config/config', './platform/dom-controller', './platform/platform', './tap-click/haptic', './navigation/deep-linker', './navigation/nav-controller', './navigation/nav-controller-base', './navigation/nav-params', './navigation/nav-util', './navigation/url-serializer', './navigation/view-controller', './components/action-sheet/action-sheet-component', './components/alert/alert-component', './components/loading/loading-component', './components/modal/modal-component', './components/picker/picker-component', './components/popover/popover-component', './components/toast/toast-component', './gestures/drag-gesture', './gestures/gesture', './gestures/slide-edge-gesture', './gestures/slide-gesture', './gestures/gesture-controller', './util/events', './util/ionic-error-handler', './platform/keyboard', './util/form', './util/util', './animations/animation', './transitions/page-transition', './transitions/transition'], factory);
    }
})(function (require, exports) {
    "use strict";
    /**
     * Import Angular
     */
    var core_1 = require('@angular/core');
    var common_1 = require('@angular/common');
    var platform_browser_1 = require('@angular/platform-browser');
    var forms_1 = require('@angular/forms');
    var http_1 = require('@angular/http');
    /**
     * Import Providers
     */
    var action_sheet_1 = require('./components/action-sheet/action-sheet');
    var alert_1 = require('./components/alert/alert');
    var app_1 = require('./components/app/app');
    var app_root_1 = require('./components/app/app-root');
    var config_1 = require('./config/config');
    var deep_linker_1 = require('./navigation/deep-linker');
    var dom_controller_1 = require('./platform/dom-controller');
    var events_1 = require('./util/events');
    var form_1 = require('./util/form');
    var gesture_controller_1 = require('./gestures/gesture-controller');
    var haptic_1 = require('./tap-click/haptic');
    var gesture_config_1 = require('./gestures/gesture-config');
    var keyboard_1 = require('./platform/keyboard');
    var loading_1 = require('./components/loading/loading');
    var menu_controller_1 = require('./components/menu/menu-controller');
    var modal_1 = require('./components/modal/modal');
    var picker_1 = require('./components/picker/picker');
    var platform_1 = require('./platform/platform');
    var platform_registry_1 = require('./platform/platform-registry');
    var popover_1 = require('./components/popover/popover');
    var tap_click_1 = require('./tap-click/tap-click');
    var toast_1 = require('./components/toast/toast');
    var mode_registry_1 = require('./config/mode-registry');
    var transition_registry_1 = require('./transitions/transition-registry');
    var transition_controller_1 = require('./transitions/transition-controller');
    var url_serializer_1 = require('./navigation/url-serializer');
    /**
     * Import Overlay Entry Components
     */
    var action_sheet_component_1 = require('./components/action-sheet/action-sheet-component');
    var alert_component_1 = require('./components/alert/alert-component');
    var app_root_2 = require('./components/app/app-root');
    var loading_component_1 = require('./components/loading/loading-component');
    var modal_component_1 = require('./components/modal/modal-component');
    var picker_component_1 = require('./components/picker/picker-component');
    var popover_component_1 = require('./components/popover/popover-component');
    var toast_component_1 = require('./components/toast/toast-component');
    /**
     * Import Components
     */
    var avatar_1 = require('./components/avatar/avatar');
    var backdrop_1 = require('./components/backdrop/backdrop');
    var badge_1 = require('./components/badge/badge');
    var button_1 = require('./components/button/button');
    var card_1 = require('./components/card/card');
    var checkbox_1 = require('./components/checkbox/checkbox');
    var chip_1 = require('./components/chip/chip');
    var click_block_1 = require('./util/click-block');
    var content_1 = require('./components/content/content');
    var datetime_1 = require('./components/datetime/datetime');
    var fab_1 = require('./components/fab/fab');
    var col_1 = require('./components/grid/col');
    var grid_1 = require('./components/grid/grid');
    var row_1 = require('./components/grid/row');
    var icon_1 = require('./components/icon/icon');
    var img_1 = require('./components/img/img');
    var infinite_scroll_1 = require('./components/infinite-scroll/infinite-scroll');
    var infinite_scroll_content_1 = require('./components/infinite-scroll/infinite-scroll-content');
    var item_1 = require('./components/item/item');
    var item_reorder_1 = require('./components/item/item-reorder');
    var item_sliding_1 = require('./components/item/item-sliding');
    var label_1 = require('./components/label/label');
    var list_1 = require('./components/list/list');
    var list_header_1 = require('./components/list/list-header');
    var menu_1 = require('./components/menu/menu');
    var menu_close_1 = require('./components/menu/menu-close');
    var menu_toggle_1 = require('./components/menu/menu-toggle');
    var native_input_1 = require('./components/input/native-input');
    var nav_1 = require('./components/nav/nav');
    var nav_pop_1 = require('./components/nav/nav-pop');
    var nav_push_1 = require('./components/nav/nav-push');
    var navbar_1 = require('./components/navbar/navbar');
    var note_1 = require('./components/note/note');
    var option_1 = require('./components/option/option');
    var overlay_portal_1 = require('./components/nav/overlay-portal');
    var picker_component_2 = require('./components/picker/picker-component');
    var radio_button_1 = require('./components/radio/radio-button');
    var radio_group_1 = require('./components/radio/radio-group');
    var range_1 = require('./components/range/range');
    var range_knob_1 = require('./components/range/range-knob');
    var refresher_1 = require('./components/refresher/refresher');
    var refresher_content_1 = require('./components/refresher/refresher-content');
    var scroll_1 = require('./components/scroll/scroll');
    var searchbar_1 = require('./components/searchbar/searchbar');
    var segment_1 = require('./components/segment/segment');
    var select_1 = require('./components/select/select');
    var show_hide_when_1 = require('./components/show-hide-when/show-hide-when');
    var slide_1 = require('./components/slides/slide');
    var slides_1 = require('./components/slides/slides');
    var spinner_1 = require('./components/spinner/spinner');
    var tab_1 = require('./components/tabs/tab');
    var tabs_1 = require('./components/tabs/tabs');
    var tab_button_1 = require('./components/tabs/tab-button');
    var tab_highlight_1 = require('./components/tabs/tab-highlight');
    var input_1 = require('./components/input/input');
    var thumbnail_1 = require('./components/thumbnail/thumbnail');
    var toggle_1 = require('./components/toggle/toggle');
    var toolbar_1 = require('./components/toolbar/toolbar');
    var toolbar_item_1 = require('./components/toolbar/toolbar-item');
    var toolbar_title_1 = require('./components/toolbar/toolbar-title');
    var typography_1 = require('./components/typography/typography');
    var virtual_scroll_1 = require('./components/virtual-scroll/virtual-scroll');
    var virtual_item_1 = require('./components/virtual-scroll/virtual-item');
    /**
     * Export Components/Directives
     */
    var action_sheet_2 = require('./components/action-sheet/action-sheet');
    exports.ActionSheet = action_sheet_2.ActionSheet;
    exports.ActionSheetController = action_sheet_2.ActionSheetController;
    var alert_2 = require('./components/alert/alert');
    exports.Alert = alert_2.Alert;
    exports.AlertController = alert_2.AlertController;
    var app_2 = require('./components/app/app');
    exports.App = app_2.App;
    var avatar_2 = require('./components/avatar/avatar');
    exports.Avatar = avatar_2.Avatar;
    var backdrop_2 = require('./components/backdrop/backdrop');
    exports.Backdrop = backdrop_2.Backdrop;
    var badge_2 = require('./components/badge/badge');
    exports.Badge = badge_2.Badge;
    var button_2 = require('./components/button/button');
    exports.Button = button_2.Button;
    var card_2 = require('./components/card/card');
    exports.Card = card_2.Card;
    exports.CardContent = card_2.CardContent;
    exports.CardHeader = card_2.CardHeader;
    exports.CardTitle = card_2.CardTitle;
    var checkbox_2 = require('./components/checkbox/checkbox');
    exports.Checkbox = checkbox_2.Checkbox;
    var chip_2 = require('./components/chip/chip');
    exports.Chip = chip_2.Chip;
    var click_block_2 = require('./util/click-block');
    exports.ClickBlock = click_block_2.ClickBlock;
    var content_2 = require('./components/content/content');
    exports.Content = content_2.Content;
    var datetime_2 = require('./components/datetime/datetime');
    exports.DateTime = datetime_2.DateTime;
    var fab_2 = require('./components/fab/fab');
    exports.FabContainer = fab_2.FabContainer;
    exports.FabButton = fab_2.FabButton;
    exports.FabList = fab_2.FabList;
    var col_2 = require('./components/grid/col');
    exports.Col = col_2.Col;
    var grid_2 = require('./components/grid/grid');
    exports.Grid = grid_2.Grid;
    var row_2 = require('./components/grid/row');
    exports.Row = row_2.Row;
    var ion_1 = require('./components/ion');
    exports.Ion = ion_1.Ion;
    var icon_2 = require('./components/icon/icon');
    exports.Icon = icon_2.Icon;
    var img_2 = require('./components/img/img');
    exports.Img = img_2.Img;
    var infinite_scroll_2 = require('./components/infinite-scroll/infinite-scroll');
    exports.InfiniteScroll = infinite_scroll_2.InfiniteScroll;
    var infinite_scroll_content_2 = require('./components/infinite-scroll/infinite-scroll-content');
    exports.InfiniteScrollContent = infinite_scroll_content_2.InfiniteScrollContent;
    var input_2 = require('./components/input/input');
    exports.TextInput = input_2.TextInput;
    var app_root_3 = require('./components/app/app-root');
    exports.IonicApp = app_root_3.IonicApp;
    var item_2 = require('./components/item/item');
    exports.Item = item_2.Item;
    exports.ItemContent = item_2.ItemContent;
    exports.ItemDivider = item_2.ItemDivider;
    exports.ItemGroup = item_2.ItemGroup;
    var item_reorder_2 = require('./components/item/item-reorder');
    exports.ItemReorder = item_reorder_2.ItemReorder;
    exports.Reorder = item_reorder_2.Reorder;
    var item_sliding_2 = require('./components/item/item-sliding');
    exports.ItemSliding = item_sliding_2.ItemSliding;
    exports.ItemOptions = item_sliding_2.ItemOptions;
    var label_2 = require('./components/label/label');
    exports.Label = label_2.Label;
    var list_2 = require('./components/list/list');
    exports.List = list_2.List;
    var list_header_2 = require('./components/list/list-header');
    exports.ListHeader = list_header_2.ListHeader;
    var loading_2 = require('./components/loading/loading');
    exports.Loading = loading_2.Loading;
    exports.LoadingController = loading_2.LoadingController;
    var menu_2 = require('./components/menu/menu');
    exports.Menu = menu_2.Menu;
    var menu_close_2 = require('./components/menu/menu-close');
    exports.MenuClose = menu_close_2.MenuClose;
    var menu_controller_2 = require('./components/menu/menu-controller');
    exports.MenuController = menu_controller_2.MenuController;
    var menu_toggle_2 = require('./components/menu/menu-toggle');
    exports.MenuToggle = menu_toggle_2.MenuToggle;
    var menu_types_1 = require('./components/menu/menu-types');
    exports.MenuType = menu_types_1.MenuType;
    var modal_2 = require('./components/modal/modal');
    exports.Modal = modal_2.Modal;
    exports.ModalController = modal_2.ModalController;
    var nav_2 = require('./components/nav/nav');
    exports.Nav = nav_2.Nav;
    var nav_pop_2 = require('./components/nav/nav-pop');
    exports.NavPop = nav_pop_2.NavPop;
    exports.NavPopAnchor = nav_pop_2.NavPopAnchor;
    var nav_push_2 = require('./components/nav/nav-push');
    exports.NavPush = nav_push_2.NavPush;
    exports.NavPushAnchor = nav_push_2.NavPushAnchor;
    var navbar_2 = require('./components/navbar/navbar');
    exports.Navbar = navbar_2.Navbar;
    var native_input_2 = require('./components/input/native-input');
    exports.NativeInput = native_input_2.NativeInput;
    exports.NextInput = native_input_2.NextInput;
    var note_2 = require('./components/note/note');
    exports.Note = note_2.Note;
    var option_2 = require('./components/option/option');
    exports.Option = option_2.Option;
    var overlay_portal_2 = require('./components/nav/overlay-portal');
    exports.OverlayPortal = overlay_portal_2.OverlayPortal;
    var picker_2 = require('./components/picker/picker');
    exports.Picker = picker_2.Picker;
    exports.PickerController = picker_2.PickerController;
    var popover_2 = require('./components/popover/popover');
    exports.Popover = popover_2.Popover;
    exports.PopoverController = popover_2.PopoverController;
    var radio_button_2 = require('./components/radio/radio-button');
    exports.RadioButton = radio_button_2.RadioButton;
    var radio_group_2 = require('./components/radio/radio-group');
    exports.RadioGroup = radio_group_2.RadioGroup;
    var range_2 = require('./components/range/range');
    exports.Range = range_2.Range;
    var range_knob_2 = require('./components/range/range-knob');
    exports.RangeKnob = range_knob_2.RangeKnob;
    var refresher_2 = require('./components/refresher/refresher');
    exports.Refresher = refresher_2.Refresher;
    var refresher_content_2 = require('./components/refresher/refresher-content');
    exports.RefresherContent = refresher_content_2.RefresherContent;
    var scroll_2 = require('./components/scroll/scroll');
    exports.Scroll = scroll_2.Scroll;
    var searchbar_2 = require('./components/searchbar/searchbar');
    exports.Searchbar = searchbar_2.Searchbar;
    var segment_2 = require('./components/segment/segment');
    exports.Segment = segment_2.Segment;
    exports.SegmentButton = segment_2.SegmentButton;
    var select_2 = require('./components/select/select');
    exports.Select = select_2.Select;
    var show_hide_when_2 = require('./components/show-hide-when/show-hide-when');
    exports.ShowWhen = show_hide_when_2.ShowWhen;
    exports.HideWhen = show_hide_when_2.HideWhen;
    exports.DisplayWhen = show_hide_when_2.DisplayWhen;
    var slide_2 = require('./components/slides/slide');
    exports.Slide = slide_2.Slide;
    var slides_2 = require('./components/slides/slides');
    exports.Slides = slides_2.Slides;
    var spinner_2 = require('./components/spinner/spinner');
    exports.Spinner = spinner_2.Spinner;
    var tab_2 = require('./components/tabs/tab');
    exports.Tab = tab_2.Tab;
    var tab_button_2 = require('./components/tabs/tab-button');
    exports.TabButton = tab_button_2.TabButton;
    var tab_highlight_2 = require('./components/tabs/tab-highlight');
    exports.TabHighlight = tab_highlight_2.TabHighlight;
    var tabs_2 = require('./components/tabs/tabs');
    exports.Tabs = tabs_2.Tabs;
    var tap_click_2 = require('./tap-click/tap-click');
    exports.TapClick = tap_click_2.TapClick;
    exports.setupTapClick = tap_click_2.setupTapClick;
    exports.isActivatable = tap_click_2.isActivatable;
    var toast_2 = require('./components/toast/toast');
    exports.Toast = toast_2.Toast;
    exports.ToastController = toast_2.ToastController;
    var toggle_2 = require('./components/toggle/toggle');
    exports.Toggle = toggle_2.Toggle;
    var toolbar_2 = require('./components/toolbar/toolbar');
    exports.Toolbar = toolbar_2.Toolbar;
    exports.ToolbarBase = toolbar_2.ToolbarBase;
    exports.Header = toolbar_2.Header;
    exports.Footer = toolbar_2.Footer;
    var toolbar_item_2 = require('./components/toolbar/toolbar-item');
    exports.ToolbarItem = toolbar_item_2.ToolbarItem;
    var toolbar_title_2 = require('./components/toolbar/toolbar-title');
    exports.ToolbarTitle = toolbar_title_2.ToolbarTitle;
    var thumbnail_2 = require('./components/thumbnail/thumbnail');
    exports.Thumbnail = thumbnail_2.Thumbnail;
    var typography_2 = require('./components/typography/typography');
    exports.Typography = typography_2.Typography;
    var virtual_scroll_2 = require('./components/virtual-scroll/virtual-scroll');
    exports.VirtualScroll = virtual_scroll_2.VirtualScroll;
    /**
     * Export Providers
     */
    var config_2 = require('./config/config');
    exports.Config = config_2.Config;
    exports.setupConfig = config_2.setupConfig;
    exports.ConfigToken = config_2.ConfigToken;
    var dom_controller_2 = require('./platform/dom-controller');
    exports.DomController = dom_controller_2.DomController;
    var platform_2 = require('./platform/platform');
    exports.Platform = platform_2.Platform;
    exports.setupPlatform = platform_2.setupPlatform;
    var haptic_2 = require('./tap-click/haptic');
    exports.Haptic = haptic_2.Haptic;
    var deep_linker_2 = require('./navigation/deep-linker');
    exports.DeepLinker = deep_linker_2.DeepLinker;
    var nav_controller_1 = require('./navigation/nav-controller');
    exports.NavController = nav_controller_1.NavController;
    var nav_controller_base_1 = require('./navigation/nav-controller-base');
    exports.NavControllerBase = nav_controller_base_1.NavControllerBase;
    var nav_params_1 = require('./navigation/nav-params');
    exports.NavParams = nav_params_1.NavParams;
    var nav_util_1 = require('./navigation/nav-util');
    exports.DeepLink = nav_util_1.DeepLink;
    exports.DeepLinkMetadata = nav_util_1.DeepLinkMetadata;
    var url_serializer_2 = require('./navigation/url-serializer');
    exports.UrlSerializer = url_serializer_2.UrlSerializer;
    exports.DeepLinkConfigToken = url_serializer_2.DeepLinkConfigToken;
    var view_controller_1 = require('./navigation/view-controller');
    exports.ViewController = view_controller_1.ViewController;
    var action_sheet_component_2 = require('./components/action-sheet/action-sheet-component');
    exports.ActionSheetCmp = action_sheet_component_2.ActionSheetCmp;
    var alert_component_2 = require('./components/alert/alert-component');
    exports.AlertCmp = alert_component_2.AlertCmp;
    var loading_component_2 = require('./components/loading/loading-component');
    exports.LoadingCmp = loading_component_2.LoadingCmp;
    var modal_component_2 = require('./components/modal/modal-component');
    exports.ModalCmp = modal_component_2.ModalCmp;
    var picker_component_3 = require('./components/picker/picker-component');
    exports.PickerCmp = picker_component_3.PickerCmp;
    exports.PickerColumnCmp = picker_component_3.PickerColumnCmp;
    var popover_component_2 = require('./components/popover/popover-component');
    exports.PopoverCmp = popover_component_2.PopoverCmp;
    var toast_component_2 = require('./components/toast/toast-component');
    exports.ToastCmp = toast_component_2.ToastCmp;
    /**
     * Export Utils
     */
    var drag_gesture_1 = require('./gestures/drag-gesture');
    exports.PanGesture = drag_gesture_1.PanGesture;
    var gesture_1 = require('./gestures/gesture');
    exports.Gesture = gesture_1.Gesture;
    var slide_edge_gesture_1 = require('./gestures/slide-edge-gesture');
    exports.SlideEdgeGesture = slide_edge_gesture_1.SlideEdgeGesture;
    var slide_gesture_1 = require('./gestures/slide-gesture');
    exports.SlideGesture = slide_gesture_1.SlideGesture;
    var gesture_controller_2 = require('./gestures/gesture-controller');
    exports.BLOCK_ALL = gesture_controller_2.BLOCK_ALL;
    exports.GESTURE_GO_BACK_SWIPE = gesture_controller_2.GESTURE_GO_BACK_SWIPE;
    exports.GESTURE_MENU_SWIPE = gesture_controller_2.GESTURE_MENU_SWIPE;
    exports.GESTURE_ITEM_SWIPE = gesture_controller_2.GESTURE_ITEM_SWIPE;
    exports.GESTURE_REFRESHER = gesture_controller_2.GESTURE_REFRESHER;
    exports.GESTURE_TOGGLE = gesture_controller_2.GESTURE_TOGGLE;
    exports.GestureController = gesture_controller_2.GestureController;
    exports.GestureDelegate = gesture_controller_2.GestureDelegate;
    exports.BlockerDelegate = gesture_controller_2.BlockerDelegate;
    var events_2 = require('./util/events');
    exports.Events = events_2.Events;
    exports.setupEvents = events_2.setupEvents;
    exports.setupProvideEvents = events_2.setupProvideEvents;
    var ionic_error_handler_1 = require('./util/ionic-error-handler');
    exports.IonicErrorHandler = ionic_error_handler_1.IonicErrorHandler;
    var keyboard_2 = require('./platform/keyboard');
    exports.Keyboard = keyboard_2.Keyboard;
    var form_2 = require('./util/form');
    exports.Form = form_2.Form;
    exports.IonicFormInput = form_2.IonicFormInput;
    exports.IonicTapInput = form_2.IonicTapInput;
    var util_1 = require('./util/util');
    exports.reorderArray = util_1.reorderArray;
    var animation_1 = require('./animations/animation');
    exports.Animation = animation_1.Animation;
    var page_transition_1 = require('./transitions/page-transition');
    exports.PageTransition = page_transition_1.PageTransition;
    var transition_1 = require('./transitions/transition');
    exports.Transition = transition_1.Transition;
    /**
     * @name IonicModule
     * @description
     * IonicModule is an NgModule that helps bootstrap a whole Ionic App. By passing a root component, IonicModule will make sure that all the components and directives from the framework are provided. This includes components such as Tabs, Menus, and Slides, as well as classes like AlertController.
     *
     *
     * We're also able to pass any configuration to our app as a second argument for `.forRoot`. This is any valid config property from [the Config Class](/docs/v2/api/config/Config/).
     *
     * The last functionality that IonicModule allows you to configure is optional routes for DeepLinker. For more information on DeepLinker, please see the [DeepLinker Docs](/docs/v2/api/navigation/DeepLinker/)
     *
     * @usage
     * ```ts
     * import { NgModule } from '@angular/core';
     * import { IonicApp, IonicModule } from 'ionic-angular';
     * import { MyApp } from './app.component';
     * import { HomePage } from '../pages/home/home';
     * @NgModule({
     *   declarations: [
     *     MyApp,
     *     HomePage
     *   ],
     *   imports: [
     *     IonicModule.forRoot(MyApp)
     *   ],
     *   bootstrap: [IonicApp],
     *   entryComponents: [
     *     MyApp,
     *     HomePage
     *   ],
     *   providers: []
     * })
     * export class AppModule {}
     * ```
     */
    var IonicModule = (function () {
        function IonicModule() {
        }
        /**
         * Set the root app component for you IonicModule
         * @param {any} appRoot The root AppComponent for this app.
         * @param {any} config Config Options for the app. Accepts any config property.
         * @param {any} deepLinkConfig Any configuration needed for the Ionic Deeplinker.
         */
        IonicModule.forRoot = function (appRoot, config, deepLinkConfig) {
            if (config === void 0) { config = null; }
            if (deepLinkConfig === void 0) { deepLinkConfig = null; }
            return {
                ngModule: IonicModule,
                providers: [
                    // useValue: bootstrap values
                    { provide: app_root_1.AppRootToken, useValue: appRoot },
                    { provide: config_1.ConfigToken, useValue: config },
                    { provide: url_serializer_1.DeepLinkConfigToken, useValue: deepLinkConfig },
                    // useFactory: user values
                    { provide: platform_registry_1.PlatformConfigToken, useFactory: platform_registry_1.providePlatformConfigs },
                    // useFactory: ionic core providers
                    { provide: platform_1.Platform, useFactory: platform_1.setupPlatform, deps: [platform_browser_1.DOCUMENT, platform_registry_1.PlatformConfigToken, core_1.NgZone] },
                    { provide: config_1.Config, useFactory: config_1.setupConfig, deps: [config_1.ConfigToken, platform_1.Platform] },
                    // useFactory: ionic app initializers
                    { provide: core_1.APP_INITIALIZER, useFactory: mode_registry_1.registerModeConfigs, deps: [config_1.Config], multi: true },
                    { provide: core_1.APP_INITIALIZER, useFactory: transition_registry_1.registerTransitions, deps: [config_1.Config], multi: true },
                    { provide: core_1.APP_INITIALIZER, useFactory: events_1.setupProvideEvents, deps: [platform_1.Platform, dom_controller_1.DomController], multi: true },
                    { provide: core_1.APP_INITIALIZER, useFactory: tap_click_1.setupTapClick, deps: [config_1.Config, platform_1.Platform, dom_controller_1.DomController, app_1.App, core_1.NgZone, gesture_controller_1.GestureController], multi: true },
                    // useClass
                    { provide: platform_browser_1.HAMMER_GESTURE_CONFIG, useClass: gesture_config_1.IonicGestureConfig },
                    // useValue
                    { provide: core_1.ANALYZE_FOR_ENTRY_COMPONENTS, useValue: appRoot, multi: true },
                    // ionic providers
                    action_sheet_1.ActionSheetController,
                    alert_1.AlertController,
                    app_1.App,
                    dom_controller_1.DomController,
                    events_1.Events,
                    form_1.Form,
                    gesture_controller_1.GestureController,
                    haptic_1.Haptic,
                    keyboard_1.Keyboard,
                    loading_1.LoadingController,
                    common_1.Location,
                    menu_controller_1.MenuController,
                    modal_1.ModalController,
                    picker_1.PickerController,
                    popover_1.PopoverController,
                    tap_click_1.TapClick,
                    toast_1.ToastController,
                    transition_controller_1.TransitionController,
                    { provide: common_1.LocationStrategy, useFactory: provideLocationStrategy, deps: [common_1.PlatformLocation, [new core_1.Inject(common_1.APP_BASE_HREF), new core_1.Optional()], config_1.Config] },
                    { provide: url_serializer_1.UrlSerializer, useFactory: url_serializer_1.setupUrlSerializer, deps: [url_serializer_1.DeepLinkConfigToken] },
                    { provide: deep_linker_1.DeepLinker, useFactory: deep_linker_1.setupDeepLinker, deps: [app_1.App, url_serializer_1.UrlSerializer, common_1.Location] },
                ]
            };
        };
        IonicModule.decorators = [
            { type: core_1.NgModule, args: [{
                        imports: [platform_browser_1.BrowserModule, http_1.HttpModule, forms_1.FormsModule, forms_1.ReactiveFormsModule],
                        exports: [
                            platform_browser_1.BrowserModule,
                            http_1.HttpModule,
                            forms_1.FormsModule,
                            forms_1.ReactiveFormsModule,
                            avatar_1.Avatar,
                            backdrop_1.Backdrop,
                            badge_1.Badge,
                            button_1.Button,
                            card_1.Card,
                            card_1.CardContent,
                            card_1.CardHeader,
                            card_1.CardTitle,
                            checkbox_1.Checkbox,
                            chip_1.Chip,
                            click_block_1.ClickBlock,
                            col_1.Col,
                            content_1.Content,
                            datetime_1.DateTime,
                            fab_1.FabContainer,
                            fab_1.FabButton,
                            fab_1.FabList,
                            toolbar_1.Footer,
                            grid_1.Grid,
                            toolbar_1.Header,
                            show_hide_when_1.HideWhen,
                            icon_1.Icon,
                            img_1.Img,
                            infinite_scroll_1.InfiniteScroll,
                            infinite_scroll_content_1.InfiniteScrollContent,
                            app_root_2.IonicApp,
                            item_1.Item,
                            item_1.ItemContent,
                            item_1.ItemDivider,
                            item_1.ItemGroup,
                            item_sliding_1.ItemOptions,
                            item_reorder_1.ItemReorder,
                            item_sliding_1.ItemSliding,
                            label_1.Label,
                            list_1.List,
                            list_header_1.ListHeader,
                            menu_1.Menu,
                            menu_close_1.MenuClose,
                            menu_toggle_1.MenuToggle,
                            native_input_1.NativeInput,
                            nav_1.Nav,
                            navbar_1.Navbar,
                            nav_pop_1.NavPop,
                            nav_pop_1.NavPopAnchor,
                            nav_push_1.NavPush,
                            nav_push_1.NavPushAnchor,
                            native_input_1.NextInput,
                            note_1.Note,
                            option_1.Option,
                            overlay_portal_1.OverlayPortal,
                            picker_component_2.PickerColumnCmp,
                            radio_button_1.RadioButton,
                            radio_group_1.RadioGroup,
                            range_1.Range,
                            range_knob_1.RangeKnob,
                            refresher_1.Refresher,
                            refresher_content_1.RefresherContent,
                            item_reorder_1.Reorder,
                            row_1.Row,
                            scroll_1.Scroll,
                            searchbar_1.Searchbar,
                            segment_1.Segment,
                            segment_1.SegmentButton,
                            select_1.Select,
                            show_hide_when_1.ShowWhen,
                            slide_1.Slide,
                            slides_1.Slides,
                            spinner_1.Spinner,
                            tab_1.Tab,
                            tabs_1.Tabs,
                            tab_button_1.TabButton,
                            tab_highlight_1.TabHighlight,
                            input_1.TextInput,
                            thumbnail_1.Thumbnail,
                            toggle_1.Toggle,
                            toolbar_1.Toolbar,
                            toolbar_item_1.ToolbarItem,
                            toolbar_title_1.ToolbarTitle,
                            typography_1.Typography,
                            virtual_item_1.VirtualFooter,
                            virtual_item_1.VirtualHeader,
                            virtual_item_1.VirtualItem,
                            virtual_scroll_1.VirtualScroll,
                        ],
                        declarations: [
                            action_sheet_component_1.ActionSheetCmp,
                            alert_component_1.AlertCmp,
                            click_block_1.ClickBlock,
                            loading_component_1.LoadingCmp,
                            modal_component_1.ModalCmp,
                            picker_component_1.PickerCmp,
                            popover_component_1.PopoverCmp,
                            toast_component_1.ToastCmp,
                            avatar_1.Avatar,
                            backdrop_1.Backdrop,
                            badge_1.Badge,
                            button_1.Button,
                            card_1.Card,
                            card_1.CardContent,
                            card_1.CardHeader,
                            card_1.CardTitle,
                            checkbox_1.Checkbox,
                            chip_1.Chip,
                            click_block_1.ClickBlock,
                            col_1.Col,
                            content_1.Content,
                            datetime_1.DateTime,
                            fab_1.FabContainer,
                            fab_1.FabButton,
                            fab_1.FabList,
                            toolbar_1.Footer,
                            grid_1.Grid,
                            toolbar_1.Header,
                            show_hide_when_1.HideWhen,
                            icon_1.Icon,
                            img_1.Img,
                            infinite_scroll_1.InfiniteScroll,
                            infinite_scroll_content_1.InfiniteScrollContent,
                            app_root_2.IonicApp,
                            item_1.Item,
                            item_1.ItemContent,
                            item_1.ItemDivider,
                            item_1.ItemGroup,
                            item_sliding_1.ItemOptions,
                            item_reorder_1.ItemReorder,
                            item_sliding_1.ItemSliding,
                            label_1.Label,
                            list_1.List,
                            list_header_1.ListHeader,
                            menu_1.Menu,
                            menu_close_1.MenuClose,
                            menu_toggle_1.MenuToggle,
                            native_input_1.NativeInput,
                            nav_1.Nav,
                            navbar_1.Navbar,
                            nav_pop_1.NavPop,
                            nav_pop_1.NavPopAnchor,
                            nav_push_1.NavPush,
                            nav_push_1.NavPushAnchor,
                            native_input_1.NextInput,
                            note_1.Note,
                            option_1.Option,
                            overlay_portal_1.OverlayPortal,
                            picker_component_2.PickerColumnCmp,
                            radio_button_1.RadioButton,
                            radio_group_1.RadioGroup,
                            range_1.Range,
                            range_knob_1.RangeKnob,
                            refresher_1.Refresher,
                            refresher_content_1.RefresherContent,
                            item_reorder_1.Reorder,
                            row_1.Row,
                            scroll_1.Scroll,
                            searchbar_1.Searchbar,
                            segment_1.Segment,
                            segment_1.SegmentButton,
                            select_1.Select,
                            show_hide_when_1.ShowWhen,
                            slide_1.Slide,
                            slides_1.Slides,
                            spinner_1.Spinner,
                            tab_1.Tab,
                            tabs_1.Tabs,
                            tab_button_1.TabButton,
                            tab_highlight_1.TabHighlight,
                            input_1.TextInput,
                            thumbnail_1.Thumbnail,
                            toggle_1.Toggle,
                            toolbar_1.Toolbar,
                            toolbar_item_1.ToolbarItem,
                            toolbar_title_1.ToolbarTitle,
                            typography_1.Typography,
                            virtual_item_1.VirtualFooter,
                            virtual_item_1.VirtualHeader,
                            virtual_item_1.VirtualItem,
                            virtual_scroll_1.VirtualScroll,
                        ],
                        entryComponents: [
                            action_sheet_component_1.ActionSheetCmp,
                            alert_component_1.AlertCmp,
                            app_root_2.IonicApp,
                            loading_component_1.LoadingCmp,
                            modal_component_1.ModalCmp,
                            picker_component_1.PickerCmp,
                            popover_component_1.PopoverCmp,
                            toast_component_1.ToastCmp
                        ]
                    },] },
        ];
        /** @nocollapse */
        IonicModule.ctorParameters = [];
        return IonicModule;
    }());
    exports.IonicModule = IonicModule;
    /**
     * @private
     */
    function provideLocationStrategy(platformLocationStrategy, baseHref, config) {
        return config.get('locationStrategy') === 'path' ?
            new common_1.PathLocationStrategy(platformLocationStrategy, baseHref) :
            new common_1.HashLocationStrategy(platformLocationStrategy, baseHref);
    }
    exports.provideLocationStrategy = provideLocationStrategy;
});
